### PR TITLE
 App Call Support.

### LINKED ISSUE #142: Add support for applications to indexer client
- [x] Update client 
- [ ] Implement steps for the integration test
- [x] Implement steps for the unit tests
- [ ] Implement function to compare response file to object

### LINKED ISSUE #160: Implement Application Transaction support
Implementation 
- [x] Implement support for `Application` Transaction type
- [x] Implement a wrapper for each of the 7 types
	- [x] Create (AppID = 0)
	- [x] Delete
	- [x] Opt In
	- [x] Opt Out
	- [x] Clear State
	- [x] Update Application
	- [x] NOP
- [x] Write an example

Testing 
- [x] Implement cucumber unitTests steps 
- [x] Implement cucumber Integreation steps

### LINKED ISSUE #185: Implement @unit.responses responses test

### PR DESCRIPTION
**Pull request by EvanJRichard**
## Summary
This PR proposes to change the Transaction class to support application calls. It also adds `makeFoo` functions for various application calls to algosdk. 

## Testing
Cucumber application tests need to pass.

### See Also
Closes #142, closes #185, and closes #160.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,8 @@
 unit:
-	node_modules/.bin/cucumber-js --tags "@unit.offline or @unit.algod or @unit.indexer or @unit.rekey" tests/cucumber/features --require tests/cucumber/steps/*
+	node_modules/.bin/cucumber-js --tags "@unit.applications" tests/cucumber/features --require tests/cucumber/steps/*
```
Comment (winder) [inline Makefile:2]: The new client tests are in `@unit.responses`

File: tests/cucumber/steps/stepdefs.js [hunk 2]
```diff
@@ -2322,3 +2344,239 @@ When('I add a rekeyTo field with the private key algorand address', function ()
...
+           default:
+               throw Error("did not recognize app arg of type" + subArg[0])
+       }
+    });
+    return appArgs;
+}
+
+When('I build an application transaction with operation {string}, application-id {int}, sender {string}, approval-program {string}, clear-program {string}, global-bytes {int}, global-ints {int}, local-bytes {int}, local-ints {int}, app-args {string}, foreign-apps {string}, app-accounts {string}, fee {int}, first-valid {int}, last-valid {int}, genesis-hash {string}', function (operationString, appIndex, sender, approvalProgramFile, clearProgramFile, numGlobalByteSlices, numGlobalInts, numLocalByteSlices, numLocalInts, appArgsCommaSeparatedString, foreignAppsCommaSeparatedString, appAccountsCommaSeparatedString, fee, firstValid, lastValid, genesisHashBase64) {
+    // operation string to enum
+    let operation = operationStringToEnum(operationString);
```
Comment (winder) [inline tests/cucumber/steps/stepdefs.js:2405]: Unfortunately, I think that each of these should use the new "makeApplicationXXX" functions. That way if you had accidentally set a wrong fields in one of those helpers the test would catch it.

File: tests/cucumber/steps/stepdefs.js [hunk 3]
```diff
@@ -2322,3 +2344,239 @@ When('I add a rekeyTo field with the private key algorand address', function ()
...
+    let fundingTxnArgs = {
+        "from": this.accounts[0],
+        "to": this.transientAddress,
+        "amount": fundingAmount,
+        "suggestedParams": sp
+    };
+    let stxKmd = await this.kcl.signTransaction(this.handle, this.wallet_pswd, fundingTxnArgs);
+    let fundingResponse = await this.v2Client.sendRawTransaction(stxKmd).do();
+    await this.v2Client.statusAfterBlock(sp["firstRound"] + 2);
+    // console.log("getting tx by id")
```
Comment (winder) [inline tests/cucumber/steps/stepdefs.js:2626]: remove old comments

**Comment by winder** [comment]
There are some new algod endpoints also (also compile/dryrun, but Pavel is working on those):
* [/v2/assets/{id}](https://github.com/justicz/go-algorand/blob/maxj/applications/daemon/algod/api/algod.oas2.json#L785)
* [/v2/applications/{id}](https://github.com/justicz/go-algorand/blob/maxj/applications/daemon/algod/api/algod.oas2.json#L722)

File: Makefile (hunk 1 above)
Comment (winder) [reply Makefile]: https://github.com/algorand/algorand-sdk-testing/blob/develop/features/unit/responses.feature

There is a new `the parsed response should equal the mock response.` that will be tricky. js/python shouldn't be too bad, but I'm a bit worried about how go will be.

File: tests/cucumber/steps/stepdefs.js [hunk 4]
```diff
@@ -1560,6 +1568,43 @@ Given('mock http responses in {string} loaded from {string}', function (fileName
...
+            this.actualMockResponse = await this.indexerClient.makeHealthCheck().do();
+        } else {
+            throw Error("did not recognize desired client \"" + client + "\"");
+        }
+    } catch (err) {
+        if (this.expectedMockResponseCode == 200) {
+            throw err;
+        }
+        if (this.expectedMockResponseCode == 500) {
+            if (!err.toString().includes("Internal Server Error")) {
```
Comment (winder) [inline tests/cucumber/steps/stepdefs.js:1595]: Is there no way to get the actual error message?

Comment (EvanJRichard) [reply tests/cucumber/steps/stepdefs.js:1595]: `err.toString()` is the actual error message, I think, unless I misunderstand?

Comment (winder) [reply tests/cucumber/steps/stepdefs.js:1595]: The error responses from the backend are designed to help the user, the error response is formatted like this:
```
{
  "data" : "oZSrqon3hT7qQqrHNUle",
  "message" : "oZSrqon3hT7qQqrHNUle"
}
```

Comment (EvanJRichard) [reply tests/cucumber/steps/stepdefs.js:1595]: Hm, unfortunately `err.toString()` just gives `Internal Server Error`, maybe we file a ticket about that.

**Pull request merged**